gradle-plugin-testing-junit/src/test/java/com/palantir/gradle/testing/integration/MavenRepoUsageTest.java

```diff
@@ -0,0 +1,167 @@
+        assertThat(result.output()).contains("com.palantir:service-a:1.0.0");
+    }
+
+    @Test
+    void publishing_same_coordinate_twice_succeeds(RootProject root, MavenRepo repo, GradleInvoker gradle) {
+        MavenCoordinate lib = MavenCoordinate.from("com.external:library:1.0.0").build();
+        repo.publish(lib);
+        repo.publish(lib);
+
+        root.buildGradle().append("""
+            dependencies {
+                implementation 'com.external:library:1.0.0'
+            }
+            """);
+
+        InvocationResult result = gradle.withArgs("dependencies").buildsSuccessfully();
+
+        assertThat(result.output()).contains("com.external:library:1.0.0");
+    }
+
+    @Test
+    void same_dependency_declared_multiple_times(RootProject root, MavenRepo repo, GradleInvoker gradle) {
+        repo.publish(MavenCoordinate.from("com.external:library:1.0.0")
+                .addDependencies("com.palantir:service-a:1.0.0")
+                .addDependencies("com.palantir:service-a:1.0.0")
+                .build());
+
+        root.buildGradle().append("""
+            dependencies {
+                implementation 'com.external:library:1.0.0'
+            }
+            """);
+
+        InvocationResult result = gradle.withArgs("dependencies").buildsSuccessfully();
+
+        assertThat(result.output()).contains("com.external:library:1.0.0");
+        assertThat(result.output()).contains("com.palantir:service-a:1.0.0");
+    }
+
+    @Test
+    void transitive_dependency_chain(RootProject root, MavenRepo repo, GradleInvoker gradle) {
+        repo.publish(
+                MavenCoordinate.from("com.external:leaf:1.0.0").build(),
+                MavenCoordinate.from("com.external:middle:1.0.0")
+                        .addDependencies("com.external:leaf:1.0.0")
+                        .build(),
+                MavenCoordinate.from("com.external:top:1.0.0")
+                        .addDependencies("com.external:middle:1.0.0")
+                        .build());
+
+        root.buildGradle().append("""
+            dependencies {
+                implementation 'com.external:top:1.0.0'
+            }
+            """);
+
+        InvocationResult result = gradle.withArgs("dependencies").buildsSuccessfully();
+
+        assertThat(result.output()).contains("com.external:top:1.0.0");
+        assertThat(result.output()).contains("com.external:middle:1.0.0");
+        assertThat(result.output()).contains("com.external:leaf:1.0.0");
+    }
+
+    @Test
+    void dependency_validation_rejects_invalid_format() {
+        assertThatThrownBy(() -> MavenCoordinate.from("com.external:library:1.0.0")
+                        .addDependencies("com.external:lib")
+                        .build())
+                .isInstanceOf(IllegalArgumentException.class)
+                .hasMessageContaining("Coordinate must be in format 'group:artifact:version'")
+                .hasMessageContaining("com.external:lib");
```

Comment:
you can just do:

```java
assertThatIllegalArgumentException()
  .isThrownBy(() -> ...)
  .hasMessageContaining("...")
  .hasMessageContaining("...");
```

I've always found the `assertThatThrownBy` clunky and doesn't really read well.

If you have an `Exception` that's not that as ubiquitous as `IllegalArgumentException` you can also do:

```java
assertThatExceptionOfType(MySpecialException.class)
  .isThrownBy(...)
  // other assertions
```